src/hierarchies/likelihoods/base_likelihood.h

```diff
@@ -0,0 +1,290 @@
+  const bayesmix::AlgorithmState::ClusterState &downcast_state(
+      const google::protobuf::Message &state_) const {
+    return google::protobuf::internal::down_cast<
+        const bayesmix::AlgorithmState::ClusterState &>(state_);
+  }
+
+  //! Current state of this cluster
+  State state;
+
+  //! Current cardinality of this cluster
+  int card = 0;
+
+  //! Logarithm of current cardinality of this cluster
+  double log_card = stan::math::NEGATIVE_INFTY;
+
+  //! Set of indexes of data points belonging to this cluster
+  std::set<int> cluster_data_idx;
+
+  //! Pointer to the cluster dataset
+  const Eigen::MatrixXd *dataset_ptr;
+};
+
+template <class Derived, typename State>
+void BaseLikelihood<Derived, State>::add_datum(
+    const int id, const Eigen::RowVectorXd &datum,
+    const Eigen::RowVectorXd &covariate) {
+  assert(cluster_data_idx.find(id) == cluster_data_idx.end());
+  set_card(++card);
+  // card += 1;
+  // log_card = std::log(card);
+  static_cast<Derived *>(this)->update_summary_statistics(datum, covariate,
+                                                          true);
+  cluster_data_idx.insert(id);
+}
+
+template <class Derived, typename State>
+void BaseLikelihood<Derived, State>::remove_datum(
+    const int id, const Eigen::RowVectorXd &datum,
+    const Eigen::RowVectorXd &covariate) {
+  static_cast<Derived *>(this)->update_summary_statistics(datum, covariate,
+                                                          false);
+  set_card(--card);
+  auto it = cluster_data_idx.find(id);
+  assert(it != cluster_data_idx.end());
+  cluster_data_idx.erase(it);
+}
+
+template <class Derived, typename State>
+void BaseLikelihood<Derived, State>::write_state_to_proto(
+    google::protobuf::Message *out) const {
+  std::shared_ptr<bayesmix::AlgorithmState::ClusterState> state_ =
+      get_state_proto();
+  auto *out_cast = downcast_state(out);
+  out_cast->CopyFrom(*state_.get());
+  out_cast->set_cardinality(card);
+}
+
+template <class Derived, typename State>
+Eigen::VectorXd BaseLikelihood<Derived, State>::lpdf_grid(
+    const Eigen::MatrixXd &data, const Eigen::MatrixXd &covariates) const {
+  Eigen::VectorXd lpdf(data.rows());
+  if (covariates.cols() == 0) {
+    // Pass null value as covariate
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived const *>(this)->lpdf(
+          data.row(i), Eigen::RowVectorXd(0));
+    }
+  } else if (covariates.rows() == 1) {
+    // Use unique covariate
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived const *>(this)->lpdf(data.row(i),
+                                                         covariates.row(0));
+    }
+  } else {
+    // Use different covariates
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived const *>(this)->lpdf(data.row(i),
+                                                         covariates.row(i));
+    }
+  }
```

Comment:
I have already seen this piece of code multiple times already, would it be possible to recycle it somehow?

Reply:
Good idea, we could do it this way

```
std::function<Eigen::RowVectorXd(int)> cov_getter;
if (covariates.cols() == 0) {
  cov_getter = [](int i){return Eigen::RowVectorXd(0)};
} else if (covariates.rows() == 1) {
  cov_getter = [&](int i){return covariates.row(0)};
} else {
  cov_getter = [&](int i){return covariates.row(i)};
}

for (int i = 0; i < data.rows(); i++) {
       lpdf(i) = static_cast<Derived const *>(this)->lpdf(
              data.row,  cov_getter(i));
 }
```

Reply:
Very nice suggestion, @brunoguindani! I will code it as @mberaha says and make all required changes.